### PR TITLE
Feature/bp/tree training

### PR DESCRIPTION
**Pull request by b-pardi**
Current implementation of tree species predictions training methods and visualizations with preliminary data reduction.

# Main features
See the model config for an overview of all tunable parameters for the pipeline. Below are some highlights and significant features:
### Image transforms
Cropped tree dimensions vary significantly. The applied transforms set a target dimension (518, 518) to get images to. If this is above a config spec'd threshold, it will downsample it to that size. It downsamples such that the longest edge is reduced to the target dimension and pads the shorter dim to fill the space, maintaining aspect ratio. If it's greater than the target dim but less than this threshold, just crop the image. If less than the target just pad the image. Note that the tree cropping script that produces these images has a min edge constraint that should prohibit images too small to be effectively upsampled.
 **NOTE** this target dimension should not be changed as this is the resolution the base model was trained on. Additionally, when dealing with dimensions, this model (and other pretrained models of similar architecture) were trained on 14x14 patches of images, so all input dimensions should be multiples of 14.

### samples per class balancing
there is a `max_class_imbalance_factor` in the config that will control how many more samples are in the class with the most samples, relative to how many are in the class with min samples. e.g. if class A has 200 samples, class B has 10 samples, and this factor is set to 10, it will limit class A to 100 samples.
Additionally there is `min_samples_per_class` with will only consider classes if they have at least that many images associated with them.

### incremental backbone layer tuning
in short, there are args in the config to control when to start allowing backbone weights to be tuned, as well as how many, and how many additional per epoch. see the model config for details.

### AMP
the training supports use of Automated Mixed Precision (AMP) which essentially uses smaller dtypes (float16) for less sensitive calculations (e.g. forward pass) and then goes back to larger dtypes for more critical calculations (e.g. fp32 for back prop).

## To run training
As long as images are located in the dir spec'd in `path_config.cropped_tree_training_images` (which is output from `1_data_prep/12_tree_crops.py`) starting training is done with 3 simple steps:

1. activating the conda env: `conda activate tsp-train`
6. adjusting any desired params in the model config found in `configs/model_config.py`
7. running `2_training/train.py`

### Notes:
- for step 2 above, one does not _need_ to change args in the config, it is sufficient to change the arg via CLI
e.g. to run for 20 epochs instead of the default 25, reduce the number of tunable backbone layers from 8, and train with data reduction (WIP), run the training script as `2_training/train.py --epochs 25 --n_last_layers_to_unfreeze 6 --use_data_reduction`

- currently the pretrained model used is the PlantCLEF2024 challenge winner, with instructions and commands given to download it in the readme. This uses the architecture `vit_base_patch14_reg4_dinov2.lvd142m` as a backbone, and is such given to timm to initialize. If other pretrained models are to be tested on different architectures, this may be added as a config arg instead.

## Things that are a work in progress
- Testing the image transforms (need to visualize some and ensure they transform as expected
- early stopping is mostly working, but due to the nature of incrementally expanding the tunable weights of the backbone, the metrics dip in performance slightly for an epoch or two, need to implement something such that early stopping does not track performance until all desired tunable weights have opened.
- Data reduction basically runs but have not confirmed it's doing what I expect, further testing soon. But the code can be read and followed to at least get an idea of what's going on.

## Summary
There's a lot more here than can be covered in one PR description. I would focus attention on the following areas to understand most of what's going on:
1. `configs/model_config.py`
2. `2_training/train.py`
3. `2_training/training_utils/step_epoch.py`
4. `2_training/training_utils/initializers.py`
5. `2_training/training_utils/data_reduction_utils.py`

I'm happy to elaborate on anything, just let me know!

### CODE REVIEW

File: 2_training/training_utils/step_epoch.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                fp[label] += (label_preds & ~label_truth).sum().cpu()
+                fn[label] += (~label_preds & label_truth).sum().cpu()
+
+            # compute the accumulated metrics
+            per_label_precision = tp.float() / (tp + fp + eps)
+            per_label_recall = tp.float() / (tp + fn + eps)
+            per_label_f1 = 2 * per_label_precision * per_label_recall / (per_label_precision + per_label_recall + eps)
+            support = (tp + fn) > 0  # ignore classes not seen so far
+
+            precision = per_label_precision[support].mean().item()
```
Comment (amrithasp02) [inline 2_training/training_utils/step_epoch.py:78]: Is my understanding correct that these metrics are being macro-averaged across the classes (i.e., giving equal importance to all classes in the final score)? It's great to have that approach, but I was just wondering if it would also be useful to report micro-averaged results so that any class imbalances do not greatly influence the final metrics.

File: 2_training/training_utils/step_epoch.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            logits = tree_model(imgs)
+            loss = criterion(logits, labels)
+
+        if training:
+            # zero previous gradients before backprop
+            optim.zero_grad()
+            if scaler is not None:
+                # scale back to fp32 and then backprop
+                scaler.scale(loss).backward()
+                scaler.step(optim) # step down the gradient
```
Comment (amrithasp02) [inline 2_training/training_utils/step_epoch.py:45]: I was looking at the [PyTorch AMP examples](https://docs.pytorch.org/docs/stable/notes/amp_examples.html) and noticed they always call `scaler.update()` after `scaler.step()`. Should we add that here as well?

File: 2_training/training_utils/step_epoch.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    # loss / acc tracking init
+    running_loss = 0.0
+    total = 0
+    correct = 0
+    num_classes = None # found in loop using model out dim
+    eps = 1e-12 # avoid div by 0
+
+    # iterate through batches
+    pbar = tqdm(dataloader, desc=pbar_msg)
+    for batch_idx, (imgs, labels, metadata) in enumerate(pbar):
```
Comment (amrithasp02) [inline 2_training/training_utils/step_epoch.py:29]: Nitpick: It looks like `metadata` doesn't get used in this `for` block. Can it be replaced with `_`?

File: 2_training/training_utils/initializers.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+    tree_dset = TreeDataset( # init dataset
+        imgs_root=path_config.cropped_tree_training_images / 'labelled',
+        gpkg_dir=path_config.drone_crowns_with_field_attributes
+    )
+    
+    # init classifier
+    tree_model = TreeSpeciesClassifierFromPretrained(
+        path_config.pretrained_model_path, # plant clef 2024, 'only_classifier_then_all' -> fine tuned backbone
+        backbone_name="vit_base_patch14_reg4_dinov2.lvd142m",
```
Comment (amrithasp02) [inline 2_training/training_utils/initializers.py:59]: Nitpick: Can the `backbone_name` also be read from the config? Just to make it easier to test with other architectures if needed in the future?

**Review by amrithasp02 — COMMENTED** [review]
The overall implementation looks really good to me! I’ve added a few comments from my initial review and from testing with the training scripts.

File: 2_training/training_utils/step_epoch.py (hunk 2 above)
Comment (b-pardi) [reply 2_training/training_utils/step_epoch.py:45]: yes good catch. Results shouldn't differ *drastically* but I will put this in the next commit.

File: 2_training/training_utils/initializers.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    for name, param in tree_model.named_parameters():
+        decay = True
+        is_backbone = name.startswith('backbone.')
+        if param.ndim < 2 or any(no_decay_type in name.lower() for no_decay_type in DECAY_EXCLUDED_PARAM_TYPES):
+            decay = False
+
+        # determine group from above dict keys -> head_ or backbone_ + decay or nodecay
+        group_key = ('backbone_' if is_backbone else 'head_') + ('decay' if decay else 'nodecay')
+        groups[group_key]['params'].append(param)
+        groups[group_key]['names'].append(param)
```
[the remark this answers is not in the document]
Comment (b-pardi) [reply 2_training/training_utils/initializers.py:44]: good catch, thank you!

File: 2_training/training_utils/step_epoch.py (hunk 1 above)
Comment (b-pardi) [reply 2_training/training_utils/step_epoch.py:78]: > Good suggestion. Also, is there any issue if some classes have zero samples or predictions that the final metrics will be `nan`?

there's no issue with zero sample classes fortunately, since the initial confusion matrix fn Amritha made showed 0s in classes that didn't have samples, so it isn't causing issues. Note though that the confusion matrix will no longer show those 0 sample classes since I added to that fn to remove classes with no samples just to clean it up.


> Is my understanding correct that these metrics are being macro-averaged across the classes (i.e., giving equal importance to all classes in the final score)? It's great to have that approach, but I was just wondering if it would also be useful to report micro-averaged results so that any class imbalances do not greatly influence the final metrics.

You are correct it is only macro avg metrics, since I had things to remedy the class imbalance with how the data is split, and we discussed synthetic minority oversampling to help imbalance as well. But by all means if you think micro avg metrics would be beneficial you're welcome to add to it :)

File: 1_data_prep/12_tree_crops.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+                for f in os.listdir(data_path):
+                    mask_fp = data_path / f # pull path of tif file
+
+                    # take the tif mask file path and change parts to get the corresponding image path
+                    path_parts = mask_fp.parts
+                    # replace mask base bath with img base path
+                    split_idx = path_parts.index(tree_label_masks.parts[-1])
+                    after = list(path_parts[split_idx+1:])
+                    #after[0] = after[0].replace('_npy', '') # remove '_npy' in dset name for imgs
+                    after[-1] = mask_fp.stem + '.JPG' # replace '.tif' file ext with img file ext
```
[the remark this answers is not in the document]
Comment (b-pardi) [reply 1_data_prep/12_tree_crops.py:73]: yup that's much simpler! I'll be honest I was stubborn for years and just using `os.path` fns and only recently started using pathlib and still learning it it seems. Thanks for catching that :)

File: 2_training/training_utils/data_utils.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+    return train_idxs, val_idxs
+
+def summarize_mixed_label_trees(dset):
+    per_tree = defaultdict(list)
+    for m in dset.meta:
+        per_tree[m['unique_treeID']].append(m['label_idx'])
+    #mixed = {tid: Counter(lbls) for tid, lbls in per_tree.items() if len(set(lbls)) > 1}
+    mixed = {}
+    for tid, lbls in per_tree.items():
+        if len(set(lbls)) > 1: # if there's multiple diff labels for one tree
```
[the remark this answers is not in the document]
Comment (b-pardi) [reply 2_training/training_utils/data_utils.py:86]: yup that's the fn that helped me figure that out lol. It will be removed in the next commits

File: 1_data_prep/12_tree_crops.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        # may also try laplacian variance and/or tenengrad score
+
+        # low res img
+        # typically trees too far off in the distance
+        if y1 - y0 < IMAGE_RES_CONSTRAINT or x1 - x0 < IMAGE_RES_CONSTRAINT:
+            continue
+
+        # check if image within safe radius to limit distortion effects
+        crop_cx, crop_cy = (x0 + x1) / 2, (y0 + y1) / 2
+        if (img_cx - crop_cx) ** 2 + (img_cy - crop_cy) ** 2 > safe_radius ** 2:
```
[the remark this answers is not in the document]
Comment (b-pardi) [reply 1_data_prep/12_tree_crops.py:139]: I updated it to be a bit more readable, however using the full norm is slower since it takes the final sqrt, this version doesn't. I updated it to show it taking the dot product of the vector of  x and y difference with itself to be more readable but maintain this more efficient approach

```
# check if image within safe radius to limit distortion effects
crop_cx, crop_cy = (x0 + x1) / 2, (y0 + y1) / 2
diff = np.array([img_cx - crop_cx, img_cy - crop_cy])
if diff @ diff > safe_radius ** 2: # similar to euclidean norm but without sqrt -> faster
    continue
```

File: 2_training/training_utils/image_processing.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    scale = target / max(w, h)
+    new_w, new_h = int(round(w * scale)), int(round(h * scale))
+    img = F.resize(img, [new_h, new_w], interpolation=InterpolationMode.BICUBIC, antialias=True)
+    pad_w = target - new_w
+    pad_h = target - new_h
+    # pad equally: (left, top, right, bottom)
+    img = F.pad(img, [pad_w // 2, pad_h // 2, pad_w - pad_w // 2, pad_h - pad_h // 2], fill=0)
+    return img
+
+def downsample_long_side_if_needed(img, long_side_thresh: int, downsample_to: int):
```
[the remark this answers is not in the document]
Comment (b-pardi) [reply 2_training/training_utils/image_processing.py:17]: > For both speed and quality, it seems helpful to only perform one resize. Can this function be combined with the one above to achieve that?

good call on combining the resizes, that will be in the next commit. as for the repeat processing in the training loop.


> Also, this type of preprocessing could be done ahead of time, even when you create the dataset, to avoid repeated computation in the training loop. But if it's running in a separate process it probably doesn't matter in practice though.

You are right that it is wasting time recomputing a lot, however my concern with doing it beforehand is that @amrithasp02 will be adding synthetic minority ovesampling, which relies on some random transforrms that _should_ be done when images are pulled to ensure it doesn't get the same random crops/resizes etc each epoch. Maybe there's a way to split this up so that these transforms are done prior to training, and there are separate transforms that rely on randomness that are done when images are pulled? Any thoughts on this?

File: 1_data_prep/12_tree_crops.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        # radial detection above should capture most of these but this exists as a failsafe
+        if img.height - y1 < IMAGE_RES_CONSTRAINT or img.width - x1 < IMAGE_RES_CONSTRAINT \
+        or y0 < IMAGE_RES_CONSTRAINT or x0 < IMAGE_RES_CONSTRAINT:
+            continue
+
+        # crop image to bbox plus padding
+        # cropping in PIL is left, top, right, bottom
+        bbox_pad_width, bbox_pad_height = (x1 - x0)  * BBOX_PADDING_RATIO, (y1 - y0) * BBOX_PADDING_RATIO
+        bbox = (x0 - bbox_pad_width, y0 - bbox_pad_height, x1 + bbox_pad_width, y1 + bbox_pad_height)
+        cropped_img = img.crop(bbox)
```
[the remark this answers is not in the document]
Comment (russelldj) [reply 1_data_prep/12_tree_crops.py:155]: I might be missing something, but my point was we're currently discarding the information between the rectangular bounding box and the minimum bounding square. This could instead be kept at this stage.
